frontend/model/contracts/group.js

```diff
@@ -774,6 +783,92 @@ sbp('chelonia/defineContract', {
         // }
       }
     },
+    'gi.contracts/group/addChatRoom': {
+      validate: objectOf({
+        chatRoomID: string,
+        attributes: objectOf(chatRoomAttributes)
+      }),
+      process ({ data, meta }, { state }) {
+        const { name, type, privacyLevel } = data.attributes
+        Vue.set(state.chatRooms, data.chatRoomID, {
+          creator: meta.username,
+          name,
+          type,
+          privacyLevel,
+          deletedDate: null,
+          users: []
+        })
+        if (!state.generalChatRoomId) {
+          Vue.set(state, 'generalChatRoomId', data.chatRoomID)
+        }
+      }
+    },
+    'gi.contracts/group/deleteChatRoom': {
+      validate: (data, { getters, meta }) => {
+        objectOf({ chatRoomID: string })(data)
+
+        if (getters.getChatRooms[data.chatRoomID].creator !== meta.username) {
+          throw new TypeError(L('Only the channel creator can delete channel.'))
+        }
+      },
+      process ({ data, meta }, { state }) {
+        Vue.delete(state.chatRooms[data.chatRoomID])
+      }
+    },
+    'gi.contracts/group/leaveChatRoom': {
+      validate: objectOf({
+        chatRoomID: string,
+        member: string,
+        leavingGroup: boolean // if kicker is exists, it means group leaving
+      }),
+      process ({ data, meta }, { state }) {
+        Vue.set(state.chatRooms[data.chatRoomID], 'users',
+          state.chatRooms[data.chatRoomID].users.filter(u => u !== data.member))
+      },
+      async sideEffect ({ meta, data }, { state }) {
+        const rootState = sbp('state/vuex/state')
+        if (meta.username === rootState.loggedIn.username && !sbp('okTurtles.data/get', 'JOINING_GROUP')) {
+          const sendingData = data.leavingGroup
+            ? { member: data.member }
+            : { member: data.member, username: meta.username }
+          await sbp('gi.actions/chatroom/leave', { contractID: data.chatRoomID, data: sendingData })
+        }
+      }
+    },
+    'gi.contracts/group/joinChatRoom': {
+      validate: objectMaybeOf({
+        username: string,
+        chatRoomID: string
+      }),
+      process ({ data, meta }, { state }) {
+        const username = data.username || meta.username
+        const { chatRoomID } = data
+        Vue.set(state.chatRooms[chatRoomID], 'users', state.chatRooms[chatRoomID].users.concat([username]))
+      },
+      async sideEffect ({ meta, data }, { state }) {
+        const rootState = sbp('state/vuex/state')
+        const username = data.username || meta.username
+        if (username === rootState.loggedIn.username) {
+          if (!sbp('okTurtles.data/get', 'JOINING_GROUP') || sbp('okTurtles.data/get', 'READY_TO_JOIN_CHATROOM')) {
+            sbp('okTurtles.data/set', 'JOINING_CHATROOM', true)
```

Comment:
A comment should be added above this line explaining why this is being set.

Reply:
Okay. Actually this is similar to setting `JOINING_GROUP` before joining group.